### PR TITLE
Improve LDAP desktop discovery

### PR DESCRIPTION
**Pull request by zmb3**
In environments with many desktops where some DNS queries time out, we can run in to situations where the discovery loop doesn't complete before the desktop's 10m TTL expires, causing desktops to disappear and reappear sporadically.

- Increase the TTL for LDAP-discovered desktops to 30m. This won't harm UX, since desktops that stop being discovered are already purged prior to their expiration.
- Perform both DNS queries (via the default resolver and by hitting the LDAP server directly) in parallel. This reduces the maximum time a single lookup can take from 20s to 5s.

### CODE REVIEW

File: lib/srv/desktop/discovery.go [hunk 1]
```diff
@@ -251,6 +280,9 @@ func (s *WindowsService) ldapEntryToWindowsDesktop(ctx context.Context, entry *l
 		return nil, trace.Wrap(err)
 	}
 
-	desktop.SetExpiry(s.cfg.Clock.Now().UTC().Add(apidefaults.ServerAnnounceTTL))
+	// We use a longer TTL for discovered desktops, because the reconciler will manually
+	// perge them if they stop being detected, and discovery of large Windows fleets can
```
Comment (Tener) [inline lib/srv/desktop/discovery.go:284]:
```suggestion
	// purge them if they stop being detected, and discovery of large Windows fleets can
```

File: lib/srv/desktop/discovery.go [hunk 2]
```diff
@@ -198,20 +194,53 @@ func (s *WindowsService) lookupDesktop(ctx context.Context, hostname string) ([]
...
+
+	// wait 5 seconds for the default resolver to return
+	select {
+	case addrs := <-defaultResult:
 		return stringAddrs(addrs), nil
+	case <-s.cfg.Clock.After(5 * time.Second):
 	}
 
-	return nil, trace.Wrap(err)
+	// If we didn't get a result from the deafault resolver,
```
Comment (Tener) [inline lib/srv/desktop/discovery.go:235]:
```suggestion
	// If we didn't get a result from the default resolver,
```

File: lib/srv/desktop/discovery.go [hunk 3]
```diff
@@ -198,20 +194,53 @@ func (s *WindowsService) lookupDesktop(ctx context.Context, hostname string) ([]
...
+			tctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+			defer cancel()
+			addrs, err := s.dnsResolver.LookupNetIP(tctx, "ip4", hostname)
+			if err != nil {
+				s.cfg.Log.Debugf("DNS lookup for %q failed with LDAP resolver: %v", hostname, err)
+			}
+			if len(addrs) > 0 {
+				ldapResult <- addrs
+			}
+		}()
```
Comment (Tener) [inline lib/srv/desktop/discovery.go:225]:
```suggestion
	const queryTimeout = time.Second * 5

	queryResolver := func(resolver *net.Resolver, resolverName string) chan []netip.Addr {
		resultChan := make(chan []netip.Addr, 1)

		if resolver != nil {
			go func() {
				tctx, cancel := context.WithTimeout(ctx, queryTimeout)
				defer cancel()
				addrs, err := resolver.LookupNetIP(tctx, "ip4", hostname)
				if err != nil {
					s.cfg.Log.Debugf("DNS lookup for %q failed with %s resolver: %v", resolverName, hostname, err)
				}
				if len(addrs) > 0 {
					resultChan <- addrs
				}
			}()
		}

		return resultChan
	}

	// kick off both DNS queries in parallel, but write results
	// to separate channels, so we can prefer the default resolver
	defaultResult := queryResolver(net.DefaultResolver, "default")
	ldapResult := queryResolver(s.dnsResolver, "LDAP")

	// wait 5 seconds for the default resolver to return
	select {
	case addrs := <-defaultResult:
		return stringAddrs(addrs), nil
	case <-s.cfg.Clock.After(queryTimeout):
	}
```
```suggestion
	// kick off both DNS queries in parallel, but write results
	// to separate channels so we can prefer the default resolver
	defaultResult := make(chan []netip.Addr, 1)
	ldapResult := make(chan []netip.Addr, 1)

	go func() {
		tctx, cancel := context.WithTimeout(ctx, 5*time.Second)
		defer cancel()
		addrs, err := net.DefaultResolver.LookupNetIP(tctx, "ip4", hostname)
		if err != nil {
			s.cfg.Log.Debugf("DNS lookup for %q failed with default resolver: %v", hostname, err)
		}
		if len(addrs) > 0 {
			defaultResult <- addrs
		}
	}()

	if s.dnsResolver != nil {
		go func() {
			tctx, cancel := context.WithTimeout(ctx, 5*time.Second)
			defer cancel()
			addrs, err := s.dnsResolver.LookupNetIP(tctx, "ip4", hostname)
			if err != nil {
				s.cfg.Log.Debugf("DNS lookup for %q failed with LDAP resolver: %v", hostname, err)
			}
			if len(addrs) > 0 {
				ldapResult <- addrs
			}
		}()
```

How do you feel about sharing the querying logic and introducing a constant for the timeout?

**Review by Tener — APPROVED** [review]
Looks good in general.

One thing to note is that we _are_ changing the effective timeouts here. Previously, the system DNS had 10 seconds of timeout, the LDPA one had... unlimited? In any case `tctx` was only used for the former, the latter used parent `ctx`. Now we trim those 10 seconds to 5, and make the LDAP query run in 5 seconds flat. (Contrast with 20 seconds we have seen in the logs today.).

I think this is a good change anyway, just highlighting this to you in case this is relevant.

**Review by ibeckermayer — APPROVED** [review]

**Review by r0mant — APPROVED** [review]

**Pull request merged**